app/stable_appconfig.go

```diff
@@ -59,7 +68,102 @@ func setCustomOrderEndBlocker() []string {
 	return []string{}
 }
 
-func (app *RegenApp) registerUpgradeHandlers() {}
+func (app *RegenApp) registerUpgradeHandlers() {
+
+	// mainnet upgrade handler
+	const upgradeName = "v4.0.0"
+	app.UpgradeKeeper.SetUpgradeHandler(upgradeName, func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		// run state migrations for sdk modules
+		toVersion, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+		if err != nil {
+			return nil, err
+		}
+
+		// run x/ecocredit state migrations
+		if err := app.smm.RunMigrations(ctx, app.AppCodec()); err != nil {
+			return nil, err
+		}
+		toVersion[ecocredit.ModuleName] = ecocreditmodule.Module{}.ConsensusVersion()
+
+		if err := recoverFunds(ctx, app.AccountKeeper, app.BankKeeper); err != nil {
+			return nil, err
+		}
+
+		return toVersion, nil
+	})
+
+	// redwood upgrade handler
+	const redwoodUpgradeName = "v4.0.0-redwood"
+	app.UpgradeKeeper.SetUpgradeHandler(upgradeName, func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
```

Comment:
As mentioned above, we should move forward with a single upgrade handler for all networks regardless of chain id and only run certain operations when chain id is either `regen-1` or `regen-redwood-1`. This upgrade handler can be removed with the changes requested above. Apologies for the confusion.